### PR TITLE
HirIdification: almost there

### PR DESCRIPTION
**Pull request by ljedrz**
The next iteration of HirIdification (#57578).

Replaces a bunch of `NodeId` method calls (mostly `as_local_node_id`) with `HirId` ones.

Removes `NodeId` from:
- [x] `PathSegment`
- [x] `PatKind`
- [x] `Destination` (replaces it with `HirId`)

In addition this PR also removes `Visitor::visit_def_mention`, which doesn't seem to be doing anything.

### CODE REVIEW

File: src/librustc/hir/map/mod.rs [hunk 1]
```diff
@@ -501,10 +501,10 @@ impl<'hir> Map<'hir> {
     }
 
     /// Given a body owner's id, returns the `BodyId` associated with it.
-    pub fn body_owned_by(&self, id: NodeId) -> BodyId {
-        self.maybe_body_owned_by(id).unwrap_or_else(|| {
-            span_bug!(self.span(id), "body_owned_by: {} has no associated body",
-                      self.node_to_string(id));
+    pub fn body_owned_by(&self, id: HirId) -> BodyId {
+        self.maybe_body_owned_by_by_hir_id(id).unwrap_or_else(|| {
```
Comment (Zoxc) [inline src/librustc/hir/map/mod.rs]: Maybe rename `maybe_body_owned_by_by_hir_id` to `maybe_body_owned_by_hir_id`? =P

Comment (ljedrz) [reply src/librustc/hir/map/mod.rs]: Yeah, that one does sound silly ^^. I'm hoping to soon be able to just remove it in favor of just `maybe_body_owned_by`.

**Comment by ljedrz** [comment]
Rebased.

**Comment by felix91gr** [comment]
I don't know where to ask this otherwise, so please forgive me if the question is out of band.

Question:

I'm seeing HirIdification all over `rustc` and `clippy`. Is `HirId` intended to replace all uses of `NodeId` eventually? I'm asking because I'm working on a Clippy lint that tries to evaluate some values in order to detect a `null` -> reference `tranmute` call. When implementing evaluation of local variables (the one case we're missing), I saw that I needed to evaluate a [`Def::Local`](https://manishearth.github.io/rust-internals-docs/rustc/hir/def/enum.Def.html#variant.Local) value, which as you can see in the link still uses `NodeId`.

Therefore, should I wait until the full `HirIdification`, or will this usage of `NodeId` keep being at the syntax-level?

Thank you so much, and sorry for disrupting your work :bowing_man:

**Comment by Zoxc** [comment]
@felix91gr Eventually all `NodeId` uses after lowering to HIR will be replaced with `HirId`. `NodeId`s will still be used before that.

You can freely convert between `NodeId` and `HirId` though with `tcx.hir().hir_to_node_id` and `tcx.hir().node_to_hir_id`, so I don't see good reason to wait for anything here. You may just have to remove these calls and update your types later on.

**Comment by felix91gr** [comment]
> Eventually all NodeId uses after lowering to HIR will be replaced with HirId. NodeIds will still be used before that.

Ah, cool, that makes perfect sense :) thanks!

> You can freely convert between NodeId and HirId though.

Ohhhhh! That's great, I didn't know that! Thank you :blush:

**Comment by ljedrz** [comment]
Added a clippy update that supersedes https://github.com/rust-lang/rust/pull/58966. Bumping priority to 1.

@bors r=Zoxc p=1

**Pull request merged**